src/java.base/share/classes/jdk/internal/foreign/Utils.java

```diff
@@ -56,28 +59,45 @@ public final class Utils {
     // Suppresses default constructor, ensuring non-instantiability.
     private Utils() {}
 
-    private static final MethodHandle BYTE_TO_BOOL;
-    private static final MethodHandle BOOL_TO_BYTE;
-    private static final MethodHandle ADDRESS_TO_LONG;
-    private static final MethodHandle LONG_TO_ADDRESS_TARGET;
-    private static final MethodHandle LONG_TO_ADDRESS_NO_TARGET;
+    private static final int
+            FILTER_ADDRESS_TO_LONG = 0,
+            FILTER_LONG_TO_POINTER = 1,
+            FILTER_LONG_TO_SEGMENT_WITH_LAYOUT = 2,
+            FILTER_LIMIT = 3;
+    private static final @Stable MethodHandle[] VH_FILTERS = new MethodHandle[FILTER_LIMIT];
+
+    private static MethodHandle filterHandle(int index) {
+        var ret = VH_FILTERS[index];
+        if (ret != null)
+            return ret;
+        return computeFilterHandle(index);
+    }
```

Comment:
Why is this using an array, instead of just having 3 fields?

Reply:
This emulates how MethodHandleImpl does the cache.

Reply:
Well, the reason I removed the eager init is that their creation in clinit is super costly. Also I think one pair of getter + creator is better than 3 pairs.

Reply:
Another possible option -- if we don't like this, is to put the constants in an holder class. As @JornVernee  says, most of the times the fields are used together. Re. initialization cost, I agree with @liach as I have observed the same thing -- these helper method handles lead to some bytecode spinning -- which is better to delay if not strictly needed?

Reply:
> I have observed the same thing -- these helper method handles lead to some bytecode spinning

Although - that was before; in the new code these helpers don't do adaptation... so it should be ok. (I also see the code has been updated)